### PR TITLE
Support for WorkerAI Provider

### PR DESCRIPTION
**Pull request by clun**
### Support for Worker AI
CloudFlare opened a Beta of a new service Called WorkerAI
https://developers.cloudflare.com/workers-ai/

They do provide a REST API for different operations.
- [Text Generation](https://developers.cloudflare.com/workers-ai/models/text-generation/)
- [Text-to-Image](https://developers.cloudflare.com/workers-ai/models/text-to-image/)
- [Text Embeddings](https://developers.cloudflare.com/workers-ai/models/text-embeddings/)- [Automatic Speech Recognition](https://developers.cloudflare.com/workers-ai/models/speech-recognition/)
- [Translation](https://developers.cloudflare.com/workers-ai/models/translation/)
- [Text Classification](https://developers.cloudflare.com/workers-ai/models/text-classification/)
- [Image Classification](https://developers.cloudflare.com/workers-ai/models/image-classification/)

This pull request proposes an implementation for `ChatModel`, `LanguageModel`, `EmbeddingModel` but also a new one `ImageModel` that can be applied for other image generation.

```java
public interface ImageModel {
    // Generate and get binary Content
    Response<byte[]> generate(String prompt);
    // Generate and save result to file
    Response<File> generate(String prompt, String destination);
```

A neat model available is Stable Diffusion XL

### CODE REVIEW

**Review by msmygit — APPROVED** [review]
Looks awesomely fantastic to me! 🚀 🎸

File: langchain4j-worker-ai/pom.xml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>dev.langchain4j</groupId>
+        <artifactId>langchain4j-parent</artifactId>
+        <version>0.24.0</version>
+        <relativePath>../langchain4j-parent/pom.xml</relativePath>
+    </parent>
+
+    <artifactId>langchain4j-worker-ai</artifactId>
```
Comment (langchain4j) [inline langchain4j-worker-ai/pom.xml]: What do you think about renaming it into `langchain4j-cloudflare-workers-ai`?

File: langchain4j-worker-ai/pom.xml [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    <packaging>jar</packaging>
+
+    <name>LangChain4j integration with CloudFlare WorkerAI</name>
```
Comment (langchain4j) [inline langchain4j-worker-ai/pom.xml]: nitpick: WorkerAI -> WorkersAI

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiChatModel.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * WorkerAI Chat model.
+ * <a href="https://developers.cloudflare.com/api/operations/workers-ai-post-run-model">...</a>
+ */
+@Slf4j
+public class WorkerAiChatModel extends AbstractWorkerAIModel implements ChatLanguageModel {
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiChatModel.java]: Please rename Worker into Workers everywhere in the code, comments, etc.

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiChatModel.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        WorkerAiChatCompletionRequest req = new WorkerAiChatCompletionRequest();
+        req.setMessages(messages.stream()
+                .map(this::toMessage)
+                .collect(Collectors.toList()));
+        return new Response<>(new AiMessage(generate(req)),null, FinishReason.STOP);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Response<AiMessage> generate(List<ChatMessage> messages, List<ToolSpecification> toolSpecifications) {
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiChatModel.java]: minor: this can be removed, default implementation takes care of throwing exception

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiChatModel.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        throw new UnsupportedOperationException("Tools are currently not supported for WorkerAI models");
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Response<AiMessage> generate(List<ChatMessage> messages, ToolSpecification toolSpecification) {
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiChatModel.java]: same here

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiChatModel.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+     * @return
+     *      text generated by the model
+     */
+    private String generate(WorkerAiChatCompletionRequest req) {
+        try {
+            retrofit2.Response<WorkerAiChatCompletionResponse> retrofitResponse = workerAiClient
+                    .generateChat(req, accountIdentifier, modelName)
+                    .execute();
+            processErrors(retrofitResponse.body(), retrofitResponse.errorBody());
+            if (retrofitResponse.body() == null) {
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiChatModel.java]: this case is already handled inside `processErrors`

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiEmbeddingModel.java [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+/**
+ * WorkerAI Embedding model.
+ * <a href="https://developers.cloudflare.com/api/operations/workers-ai-post-run-model">...</a>
+ */
+@Slf4j
+public class WorkerAiEmbeddingModel extends AbstractWorkerAIModel implements EmbeddingModel {
+
+    /**
+     * Constructor with Builder.
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiEmbeddingModel.java]: I am not sure that many javadocs in this PR are useful, most of them do not contribute any new information

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiEmbeddingModel.java [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        try {
+            WorkerAiEmbeddingRequest req = new WorkerAiEmbeddingRequest();
+            req.getText().add(text);
+
+            retrofit2.Response<WorkerAiEmbeddingResponse> retrofitResponse = workerAiClient
+                    .embed(req, accountIdentifier, modelName)
+                    .execute();
+
+            processErrors(retrofitResponse.body(), retrofitResponse.errorBody());
+            if (retrofitResponse.body() == null) {
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiEmbeddingModel.java]: same here

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiEmbeddingModel.java [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+                throw new RuntimeException("Unexpected response: " + retrofitResponse);
+            }
+            WorkerAiEmbeddingResponse.EmbeddingResult res = retrofitResponse.body().getResult();
+            // Single Vector expected
+            if (res.getShape().get(0) != 1) {
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiEmbeddingModel.java]: what about moving this into `res.ensureSingleVector()` ?

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiEmbeddingModel.java [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+                throw new RuntimeException("Unexpected shape: " + res.getShape());
+            }
+            List<Float> embeddings = res.getData().get(0);
+            float[] floatArray = new float[embeddings.size()];
+            for (int i = 0; i < embeddings.size(); i++) {
+                floatArray[i] = embeddings.get(i); // Unboxing Float to float
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiEmbeddingModel.java]: you could use `Embedding.from(embeddings)` instead

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiEmbeddingModel.java [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+            throw new RuntimeException("Unexpected response: " + retrofitResponse);
+        }
+        WorkerAiEmbeddingResponse.EmbeddingResult res = retrofitResponse.body().getResult();
+
+        List<List<Float>> embeddings = res.getData();
+        List<Embedding> embeddingsList = new ArrayList<>();
+        for (List<Float> embedding : embeddings) {
+            float[] floatArray = new float[embedding.size()];
+            for (int i = 0; i < embedding.size(); i++) {
+                floatArray[i] = embedding.get(i); // Unboxing Float to float
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiEmbeddingModel.java]: same here

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiEmbeddingModel.java [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+    throws IOException {
+        WorkerAiEmbeddingRequest req = new WorkerAiEmbeddingRequest();
+        for (TextSegment textSegment : chunk) {
+            req.getText().add(textSegment.text());
+        }
+        retrofit2.Response<WorkerAiEmbeddingResponse> retrofitResponse = workerAiClient
+                .embed(req, accountIdentifier, modelName)
+                .execute();
+        processErrors(retrofitResponse.body(), retrofitResponse.errorBody());
+        if (retrofitResponse.body() == null) {
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiEmbeddingModel.java]: same here

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiEmbeddingModel.java [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+
+    /** {@inheritDoc} */
+    @Override
+    public Response<List<Embedding>> embedAll(List<TextSegment> textSegments) {
+        List<Future<List<Embedding>>> futures = new ArrayList<>();
+        ExecutorService executor = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors());
+        try {
+            final int chunkSize = 100;
+            for (int i = 0; i < textSegments.size(); i += chunkSize) {
+                List<TextSegment> chunk = textSegments.subList(i, Math.min(textSegments.size(), i + chunkSize));
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiEmbeddingModel.java]: How about moving partitioning logic into `Util.partition(List<?> list, int partitionSize)`? Would also be great to have a test for a few scenarios (null, 0, 1, 99, 100, 101, 199, 200, 201 items, etc)

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/model/ImageModel.java [hunk 14]
```diff
@@ -0,0 +1,34 @@
+package dev.langchain4j.model.workerai.model;
+
+import dev.langchain4j.model.output.Response;
+
+import java.io.File;
+
+/**
+ * Model to work with images. It can generate or classify
+ */
+public interface ImageModel {
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/model/ImageModel.java]: Please move `ImageModel` into `langchain4j-core` module, `dev.langchain4j.model.image` package, this will be also used by DALL-E integration by @Heezer

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/model/ImageModel.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+
+    /**
+     * Generate an image and get its binary content.
+     *
+     * @param prompt
+     *       image prompt
+     * @return
+     *      generated image content
+     */
+    Response<byte[]> generate(String prompt);
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/model/ImageModel.java]: I think it would be better to return `Response<Image>` here and `Image` will encapsulate `byte[]` inside (for other cases it can be also `url` of the image). But the idea is the same, when you call `generate()` on an `ImageModel`, `Image` is returned, regardless of the format

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/model/ImageModel.java [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+     * Generate an image and save it to the destination.
+     *
+     * @param prompt
+     *       image prompt
+     * @param destination
+     *      image destination
+     * @return
+     *      generated image
+     */
+    Response<File> generate(String prompt, String destination);
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/model/ImageModel.java]: How about:
`Response<Void> generate(String prompt, Path destination);` <- assuming user does not need to manipulate the image in the code, he just wants to generate and store it locally. And even if he wants to manipulate it, he has a `Path` which he provided himself already, so he can use it directly
or
`Response<Image> generate(String prompt, Path destination);` <- gives a bit more flexibility: e.g. returning an `Image` as a url (e.g. for sending it to FE for immediate display), but still persisting it into destination

File: langchain4j-worker-ai/src/test/java/dev/langchain4j/model/workerai/WorkerAIEmbeddingModelIT.java [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+                .modelName(WorkerAiModelName.BAAI_EMBEDDING_BASE)
+                .accountIdentifier(System.getenv("WORKERAI_ACCOUNT_ID"))
+                .token(System.getenv("WORKERAI_API_KEY"))
+                .buildEmbeddingModel();
+    }
+
+    @Test
+    public void generateEmbeddingSimple() {
+        Response<Embedding> out = embeddingModel.embed("Sentence1");
+        Assertions.assertNotNull(out.content());
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/test/java/dev/langchain4j/model/workerai/WorkerAIEmbeddingModelIT.java]: I would also add `assertThat(out.content().dimension()).isEqualTo(768);`

File: langchain4j-worker-ai/src/test/java/dev/langchain4j/model/workerai/WorkerAIEmbeddingModelIT.java [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    @Test
+    public void generateEmbeddings() {
+        List<TextSegment> data = new ArrayList<>();
+        data.add(new TextSegment("Sentence1", new Metadata()));
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/test/java/dev/langchain4j/model/workerai/WorkerAIEmbeddingModelIT.java]: nitpick: you could use `TextSegment.from("Sentence1")`

File: langchain4j-worker-ai/src/test/java/dev/langchain4j/model/workerai/WorkerAIEmbeddingModelIT.java [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+        data.add(new TextSegment("Sentence2", new Metadata()));
+        Response<List<Embedding>> out = embeddingModel.embedAll(data);
+        Assertions.assertNotNull(out.content());
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/test/java/dev/langchain4j/model/workerai/WorkerAIEmbeddingModelIT.java]: please add a check that there are 2 embeddings in the response.
Ideally you could embed "hi" and "hello" and then do `assertThat(CosineSimilarity.between(out.content().get(0), out.content().get(1))).isBetween(0.9, 1.0);`

File: langchain4j-worker-ai/src/test/java/dev/langchain4j/model/workerai/WorkerAILanguageModelIT.java [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+    public static void initializeModel() {
+        languageModel = WorkerAiChatModel.builder()
+                .modelName(WorkerAiModelName.LLAMA2_7B_FULL)
+                .accountIdentifier(System.getenv("WORKERAI_ACCOUNT_ID"))
+                .token(System.getenv("WORKERAI_API_KEY"))
+                .buildLanguageModel();
+    }
+    @Test
+    public void generateText() {
+        Response<String> joke = languageModel.generate("Tell me jokeC cloud");
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/test/java/dev/langchain4j/model/workerai/WorkerAILanguageModelIT.java]: nitpick: some corrupted text

File: langchain4j-worker-ai/src/test/java/dev/langchain4j/model/workerai/WorkerAILanguageModelIT.java [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+        Assertions.assertNotNull(joke);
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/test/java/dev/langchain4j/model/workerai/WorkerAILanguageModelIT.java]: please add a check that joke.content() is not empty

File: langchain4j-worker-ai/src/test/java/dev/langchain4j/model/workerai/WorkerAiImageModelIT.java [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+    @Test
+    void should_generate_an_image_as_binary() {
+        Response<byte[]> image = imageModel.generate("Draw me a squirrel");;
+        Assertions.assertNotNull(image.content());
+        Assertions.assertTrue(image.content().length > 0);
+    }
+
+    @Test
+    void should_generate_an_image_as_file() {
+        String homeDirectory = System.getProperty("user.home");
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/test/java/dev/langchain4j/model/workerai/WorkerAiImageModelIT.java]: please use [JUnit's temp directory](https://www.baeldung.com/junit-5-temporary-directory) instead

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/client/AbstractWorkerAIModel.java [hunk 23]
```diff
@@ -0,0 +1,205 @@
...
+import okhttp3.ResponseBody;
+
+import java.io.IOException;
+
+/**
+ * Abstract class for WorkerAI models as they are all initialized the same way.
+ * <a href="https://developers.cloudflare.com/api/operations/workers-ai-post-run-model">...</a>
+ */
+@Slf4j
+public abstract class AbstractWorkerAIModel {
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/client/AbstractWorkerAIModel.java]: Would be awesome to allow user define `Duration timeout` and `Integer maxRetries` with sensible defaults (see `OpenAiChatModel` as example), but this can be added later

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiLanguageModel.java [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+
+    /** {@inheritDoc} */
+    @Override
+    public Response<String> generate(String prompt) {
+        try {
+            retrofit2.Response<WorkerAiTextCompletionResponse> retrofitResponse = workerAiClient
+                    .generateText(new WorkerAiTextCompletionRequest(prompt), accountIdentifier, modelName)
+                    .execute();
+            processErrors(retrofitResponse.body(), retrofitResponse.errorBody());
+            if (retrofitResponse.body() == null) {
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiLanguageModel.java]: same here

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiLanguageModel.java [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+            }
+            return new Response<>(retrofitResponse.body().getResult().getResponse());
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Response<String> generate(Prompt prompt) {
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiLanguageModel.java]: why do you need to override default impl?

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiChatModel.java [hunk 26]
```diff
@@ -0,0 +1,109 @@
...
+
+    /** {@inheritDoc} */
+    @Override
+    public String generate(String userMessage) {
+        return generate(new WorkerAiChatCompletionRequest(WorkerAiChatCompletionRequest.MessageRole.user, userMessage));
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Response<AiMessage> generate(@NonNull  ChatMessage... messages) {
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiChatModel.java]: Why do you need to override the default implementation from `ChatLanguageModel`?

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiEmbeddingModel.java [hunk 27]
```diff
@@ -0,0 +1,165 @@
...
+     *      api token from .
+     */
+    @SuppressWarnings("unused")
+    public WorkerAiEmbeddingModel(String accountIdentifier, String modelName, String token) {
+        super(accountIdentifier, modelName, token);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Response<Embedding> embed(String text) {
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiEmbeddingModel.java]: it seems that instead of duplicating the logic of `Response<List<Embedding>> embedAll(List<TextSegment> textSegments)` you could simply remove this and `Response<Embedding> embed(TextSegment textSegment)` methods and the default implementation will call `Response<List<Embedding>> embedAll(List<TextSegment> textSegments)` eventually

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/client/AbstractWorkerAIModel.java [hunk 28]
```diff
@@ -0,0 +1,205 @@
...
+     * @param accountIdentifier
+     *      account identifier.
+     * @param modelName
+     *      model name.
+     * @param token
+     *      api token from .
+     */
+    public AbstractWorkerAIModel(String accountIdentifier, String modelName, String token) {
+        if (accountIdentifier == null || accountIdentifier.isEmpty()) {
+            throw new IllegalArgumentException("Account identifier should not be null or empty");
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/client/AbstractWorkerAIModel.java]: nitpick: you could do `this.accountIdentifier = ensureNotBlank(accountIdentifier, "accountIdentifier")` instead

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/client/AbstractWorkerAIModel.java [hunk 29]
```diff
@@ -0,0 +1,205 @@
...
+     *      response
+     * @param errors
+     *      errors body from retrofit
+     * @throws IOException
+     *      error occurred during invocation
+     */
+    protected void processErrors(ApiResponse<?> res, ResponseBody errors)
+    throws IOException {
+        if (res == null || !res.isSuccess()) {
+            StringBuilder errorMessage = new StringBuilder("Failed to generate chat message:");
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/client/AbstractWorkerAIModel.java]: this method is used not only by chat model, so better to generalize the error message

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/client/AbstractWorkerAIModel.java [hunk 30]
```diff
@@ -0,0 +1,205 @@
...
+            if (res == null) {
+                errorMessage.append(errors.string());
+            } else if (res.getErrors() != null) {
+                errorMessage.append(res.getErrors().stream()
+                        .map(ApiResponse.Error::getMessage)
+                        .reduce((a, b) -> a + "\n" + b)
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/client/AbstractWorkerAIModel.java]: minor: this can be simplified into `.collect(joining("\n"))`

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/client/AbstractWorkerAIModel.java [hunk 31]
```diff
@@ -0,0 +1,205 @@
...
+                        .orElse(""));
+            }
+            log.error(errorMessage.toString());
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/client/AbstractWorkerAIModel.java]: why is it needed to both log and throw the same exception message?

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/client/WorkerAiImageGenerationRequest.java [hunk 32]
```diff
@@ -0,0 +1,27 @@
...
+
+    /**
+     * Prompt to generate the image.
+     */
+    String prompt;
+
+    /**
+     * File to save the image.
+     */
+    String destinationFile;
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/client/WorkerAiImageGenerationRequest.java]: I guess `destinationFile` should be removed from this class

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/client/WorkerAiImageGenerationResponse.java [hunk 33]
```diff
@@ -0,0 +1,36 @@
...
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+import java.io.InputStream;
+
+/**
+ * Response to generate an image.
+ */
+public class WorkerAiImageGenerationResponse
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/client/WorkerAiImageGenerationResponse.java]: it seems that this class is not used anywhere, can be removed

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiChatModel.java [hunk 34]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Response<AiMessage> generate(List<ChatMessage> messages) {
+        WorkerAiChatCompletionRequest req = new WorkerAiChatCompletionRequest();
+        req.setMessages(messages.stream()
+                .map(this::toMessage)
+                .collect(Collectors.toList()));
+        return new Response<>(new AiMessage(generate(req)),null, FinishReason.STOP);
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiChatModel.java]: If WorkersAI do not provide a finish reason, I would better return null instead of STOP as it might not always be true.
Same for all other places where you return STOP.

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiImageModel.java [hunk 35]
```diff
@@ -0,0 +1,88 @@
...
+            // Mapping inbound
+            WorkerAiImageGenerationRequest imgReq = new WorkerAiImageGenerationRequest();
+            imgReq.setPrompt(prompt);
+
+            retrofit2.Response<ResponseBody> response = workerAiClient
+                    .generateImage(imgReq, accountIdentifier, modelName)
+                    .execute();
+
+            if (response.isSuccessful() && response.body() != null) {
+                InputStream inputStream = response.body().byteStream();
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiImageModel.java]: minor: the code inside if block can be extracted into a separate method

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiModelName.java [hunk 36]
```diff
@@ -0,0 +1,73 @@
...
+ * Group model name in a class to ease the usage.
+ * It has been preferred as an Enum for flexibility.
+ */
+@SuppressWarnings("unused")
+public class WorkerAiModelName {
+
+    /**
+     * Max token for text generation
+     */
+    public static final int MAX_TOKENS =  512;
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiModelName.java]: Is this used somewhere?

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiModelName.java [hunk 37]
```diff
@@ -0,0 +1,73 @@
...
+     */
+    public static final String LLAMA2_7B_FULL = "@cf/meta/llama-2-7b-chat-fp16";
+    /**
+     * Model names
+     */
+    public static final String LLAMA2_7B_QUANTIZED = "@cf/meta/llama-2-7b-chat-int8";
+    /**
+     * Model names
+     */
+    public static final String LLAMA2_7B_AWQ = "@hf/thebloke/codellama-7b-instruct-awq";
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiModelName.java]: is it supposed to be llama or codellama?

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiModelName.java [hunk 38]
```diff
@@ -0,0 +1,73 @@
...
+    // ---------------------------------------------------------------------
+    // Text Embeddings
+    // https://developers.cloudflare.com/workers-ai/models/text-embeddings/
+    // ---------------------------------------------------------------------
+
+    /**
+     * Model names
+     * maxToken: 512, dimension 384
+     */
+    public static final String BAAI_EMBEDDING_SMALL = "@cf/baai/bge-small-en-v1.5";
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/WorkerAiModelName.java]: minor: I would include versions of the models in the constant names as there is a high chance that another version will come soon

File: langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/client/AbstractWorkerAIModel.java [hunk 39]
```diff
@@ -0,0 +1,205 @@
...
+     * ModelName, preferred as enum for extensibility.
+     *
+     * @see dev.langchain4j.model.workerai.WorkerAiModelName
+     */
+    protected String modelName;
+
+    /**
+     * OkHttpClient for the WorkerAI API.
+     *
+     * @see dev.langchain4j.model.workerai.WorkerAiModelName
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/main/java/dev/langchain4j/model/workerai/client/AbstractWorkerAIModel.java]: copy-paste

File: langchain4j-worker-ai/src/test/java/dev/langchain4j/model/workerai/WorkerAIChatModelIT.java [hunk 40]
```diff
@@ -0,0 +1,74 @@
...
+        List<ChatMessage> conversation = new ArrayList<>();
+        conversation.add(systemMessage("You an an assistant i will give you the name of " +
+                "a country and you will give me exactly the name of the capital, " +
+                "no other text or message, " +
+                "just the name of the city"));
+        conversation.add(userMessage("France"));
+        Response<AiMessage> response = chatModel.generate(conversation);
+        Assertions.assertNotNull(response);
+        assertThat(response.content().text()).isNotBlank();
+        Assertions.assertEquals("PARIS", chatModel.generate(conversation).content().text().toUpperCase());
```
Comment (langchain4j) [inline langchain4j-worker-ai/src/test/java/dev/langchain4j/model/workerai/WorkerAIChatModelIT.java]: nitpick: you could do `assertThat(response.content().text()).isEqualToIgnoringCase("Paris")` instead

**Review by ghost — CHANGES_REQUESTED** [review]
@clun thank you a lot, you rock! I think noone added so much functionality at once yet 🚀

**Comment by geoand** [comment]
Any chance this can be rebooted?

**Comment by clun** [comment]
Sure the lagnchain4j providers interfaces evolved quite a bit. 
I am using this code in production for a little while. Will give a try this week-end.

**Comment by langchain4j** [comment]
>> lagnchain4j providers interfaces evolved quite a bit

Hmm... what interfaces?

**Comment by clun** [comment]
The image model did not exist at that time, I also had to find a way to support image editions and masks.
Here is the new tentative @geoand 
https://github.com/langchain4j/langchain4j/pull/1262
